File: modules/coupon-domain/src/main/java/com/thanh0x/coursedeal/model/coupon/CourseLevelConverter.java
package com.thanh0x.coursedeal.model.coupon;

import jakarta.persistence.AttributeConverter;
import jakarta.persistence.Converter;

/**
 * JPA Converter to handle flexible string-to-enum mapping for CourseLevel.
 * This prevents 500 errors when the database contains strings that don't exactly match enum names.
 */
@Converter(autoApply = true)
public class CourseLevelConverter implements AttributeConverter<CourseLevel, String> {

    @Override
    public String convertToDatabaseColumn(CourseLevel level) {
        if (level == null) return null;
        return level.name();
    }

    @Override
    public CourseLevel convertToEntityAttribute(String dbData) {
        if (dbData == null || dbData.isBlank()) return CourseLevel.ALL_LEVELS;
        return CourseLevel.fromString(dbData);
    }
}

File: modules/coupon-domain/src/main/kotlin/com/thanh0x/coursedeal/model/coupon/CourseLevelConverter.kt
package com.thanh0x.coursedeal.model.coupon

import jakarta.persistence.AttributeConverter
import jakarta.persistence.Converter

/**
 * JPA Converter to handle flexible string-to-enum mapping for CourseLevel.
 * This prevents 500 errors when the database contains strings that don't exactly match enum names.
 */
@Converter(autoApply = true)
class CourseLevelConverter : AttributeConverter<CourseLevel, String> {

    override fun convertToDatabaseColumn(level: CourseLevel?): String? {
        return level?.name
    }

    override fun convertToEntityAttribute(dbData: String?): CourseLevel {
        if (dbData.isNullOrBlank()) return CourseLevel.ALL_LEVELS
        return CourseLevel.fromString(dbData)
    }
}
